integration/pkcs11/pkcs11_test.go

```diff
@@ -0,0 +1,420 @@
+// +build pkcs11
+
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+package pkcs11
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/hex"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	bpkcs11 "github.com/hyperledger/fabric/bccsp/pkcs11"
+
+	"github.com/miekg/pkcs11"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/hyperledger/fabric/integration/nwo/fabricconfig"
+	"github.com/tedsuo/ifrit"
+
+	"github.com/hyperledger/fabric/integration/nwo"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Configures PKCS#11 for peer and orderer", func() {
+	var (
+		tempDir string
+		network *nwo.Network
+		client  *docker.Client
+		process ifrit.Process
+		peer    *nwo.Peer
+		orderer *nwo.Orderer
+	)
+
+	BeforeEach(func() {
+		var err error
+		tempDir, err = ioutil.TempDir("", "p11")
+		Expect(err).NotTo(HaveOccurred())
+
+		client, err = docker.NewClientFromEnv()
+		Expect(err).NotTo(HaveOccurred())
+
+		network = nwo.New(nwo.BasicSolo(), tempDir, client, StartPort(), components)
+		network.GenerateConfigTree()
+
+		peer = network.Peers[0]
+		orderer = network.Orderer("orderer")
+
+		lib, pin, label := bpkcs11.FindPKCS11Lib()
+		myBCCSP := &fabricconfig.BCCSP{
+			Default: "PKCS11",
+			PKCS11: &fabricconfig.PKCS11{
+				Security: 256,
+				Hash:     "SHA2",
+				Pin:      pin,
+				Label:    label,
+				Library:  lib,
+			},
+		}
+
+		By("update bccsp peer config")
+		peerConfig := network.ReadPeerConfig(peer)
+		peerConfig.Peer.BCCSP = myBCCSP
+		network.WritePeerConfig(peer, peerConfig)
+
+		By("Update bccsp orderer config")
+		ordererConfig := network.ReadOrdererConfig(orderer)
+		ordererConfig.General.BCCSP = myBCCSP
+		network.WriteOrdererConfig(orderer, ordererConfig)
+
+		network.Bootstrap()
+
+		ctx, sess, err := setupPKCS11Ctx(lib, label, pin)
+		Expect(err).NotTo(HaveOccurred())
+
+		configurePeerPKCS11(ctx, sess, network)
+		configureOrdererPKCS11(ctx, sess, network)
+
+		ctx.CloseSession(sess)
```

Comment:
This should probably be a defer because of the assertions happening out of the config functions.